### PR TITLE
[v1.8.x] Automate website artifacts uploading

### PR DESCRIPTION
**Pull request by waytrue17**
All versions of MXNet website artifacts are compressed into a zip file and stored in a private S3 bucket, with the following file structure:
> version
> ----1.0.0
> ----1.1.0
>...
> ----1.6.0
> ----1.7.0

For new MXNet versions released, we need to add the new website artifacts into the above zip file and upload it to the S3 bucket. Currently we have a pipeline that generates version specific artifacts, but it's still a manual process to upload them to s3 in the expected format shown above. 
This PR aims to automate the website artifact uploading process.
We introduced pipeline `Jenkinsfile_website_upload` to build the new version website artifacts, prepare all versions of website artifacts into the expected format, and upload them to the target S3 bucket. This pipeline utilizes Jenkins environment variable `FOLDER_NAME` to specify the name of the new artifacts to be added.

### CODE REVIEW

File: ci/docker/runtime_functions.sh
Comment (mseth10) [inline ci/docker/runtime_functions.sh:1933]: check if version name folder already exists, if it does delete it

File: ci/docker/runtime_functions.sh
Comment (mseth10) [inline ci/docker/runtime_functions.sh:1934]: change zip file name

File: ci/jenkins/Jenkins_steps.groovy
Comment (mseth10) [inline ci/jenkins/Jenkins_steps.groovy:1612]: we dont need this

**Comment by mseth10** [comment]
After this PR gets merged, modify Step 3.2.1 Point 5 in
https://cwiki.apache.org/confluence/display/MXNET/Release+Process#ReleaseProcess-3.2.UpdatetheMXNetwebsite

File: ci/docker/runtime_functions.sh
Comment (waytrue17) [reply ci/docker/runtime_functions.sh]: Done

File: ci/docker/runtime_functions.sh
Comment (waytrue17) [reply ci/docker/runtime_functions.sh]: Done

**Review by mseth10 — APPROVED** [review]
LGTM

**Comment by mseth10** [comment]
The PR changes are being tested by this Jenkins pipeline: http://jenkins.mxnet-ci-dev.amazon-ml.com/job/docs/job/sethman-website-artifacts/

**Comment by waytrue17** [comment]
Hi @ys2843 Could you please review this PR? Thanks!

**Review by ys2843 — COMMENTED** [review]
Great work! Just one question:

File: ci/docker/runtime_functions.sh
Comment (waytrue17) [reply ci/docker/runtime_functions.sh:1942]: Yes, It should be accessible since I passed the `bucket-owner-full-control` to the s3 bucket. The uploaded file should have the same permissions as other files in the bucket.

File: ci/jenkins/Jenkinsfile_website_release
Comment (marcoabreu) [inline ci/jenkins/Jenkinsfile_website_release:33]: Since this is a release pipeline, it must be run as restricted job on restricted slaves. Regular slaves are not permitted to have any publish access.

@leezu @aaronmarkham

**Review by marcoabreu — CHANGES_REQUESTED** [review]

File: ci/jenkins/Jenkinsfile_website_release
Comment (mseth10) [reply ci/jenkins/Jenkinsfile_website_release:33]: @marcoabreu Just to clarify, this pipeline is more of a utility pipeline that is run manually to build artifacts corresponding to a specific version and package them in s3 zip file. It does not publish anything to the MXNet website. It automates the job that is otherwise required to be done manually by the release manager.
It is an improvement over the pipeline that was used to build v1.7 artifacts: http://jenkins.mxnet-ci-dev.amazon-ml.com/job/docs/job/1.7-test/ which used regular slaves as well. This PR aims to check-in the pipeline code and add a stage to package artifacts in s3 zip file.

File: ci/docker/runtime_functions.sh
Comment (mseth10) [reply ci/docker/runtime_functions.sh]: We do not want to change the s3 bucket accessibility permissions with this PR. @ys2843 can you please confirm that `public-read` is the right permission for the bucket?

File: ci/jenkins/Jenkinsfile_website_release
Comment (waytrue17) [reply ci/jenkins/Jenkinsfile_website_release]: @marcoabreu Just like @mseth10 mentioned, this pipeline will only be used to build the new website artifacts and upload them to a private S3 bucket for internal use. Sorry that the file name `Jenkinsfile_website_release` might be confusing. I've changed it to `Jenkinsfile_website_upload`.

Comment (marcoabreu) [reply ci/jenkins/Jenkinsfile_website_release]: None the less, you are giving the pipeline write permissions. Our unrestricted slaves are set up to specifically not be able to persist anything. 

Otherwise, somebody could go ahead, submit a PR which hijacks a pipeline. They could then go ahead and download illegal content from somewhere and host it in our s3 bucket. Do you want to be responsible for that?

Back then, the slaves were specifically set up with an instance profile which prohibited all these actions. You might want to engage in a security review to ensure everything is still safe.

Comment (mseth10) [reply ci/jenkins/Jenkinsfile_website_release]: > None the less, you are giving the pipeline write permissions. Our unrestricted slaves are set up to specifically not be able to persist anything.

Agreed, we can move the pipeline to jenkins-ci prod account, where we have access to restricted nodes.

Comment (waytrue17) [reply ci/jenkins/Jenkinsfile_website_release]: @marcoabrew @mseth10 The pipeline has beed switched to restricted node and tested on jenkins-ci prod account.

**Comment by waytrue17** [comment]
@mxnet-bot run ci [unix-cpu]

File: ci/jenkins/Jenkins_steps.groovy
Comment (marcoabreu) [inline ci/jenkins/Jenkins_steps.groovy:1612]: Try to avoid archive artifacts as much as possible as the space on the jenkins master is limited. Instead, use stash to pass between stages and s3 to pass between pipelines. Same below

Comment (mseth10) [reply ci/jenkins/Jenkins_steps.groovy:1612]: @waytrue17 we can skip archiving `full_website.tgz` as it is already stashed

File: ci/jenkins/Jenkins_steps.groovy
Comment (mseth10) [inline ci/jenkins/Jenkins_steps.groovy:1630]:
```suggestion
              utils.unpack_and_init('full_website', 'docs/_build/full_website.tgz')
```

@waytrue17 we can also make use of this utility fn here.

File: ci/jenkins/Jenkins_steps.groovy
Comment (waytrue17) [reply ci/jenkins/Jenkins_steps.groovy]: removed archiving call

File: ci/jenkins/Jenkins_steps.groovy
Comment (waytrue17) [reply ci/jenkins/Jenkins_steps.groovy]: switched to `utils.unpack_and_init` and tested it on prod account

**Comment by samskalicky** [comment]
@marcoabreu can you please review again now that your feedback has been addressed?

**Comment by sandeep-krishnamurthy** [comment]
@marcoabreu - Looks like your review comments are addressed. Can you take a look at this and approve / have other feedback?

**Review by ys2843 — APPROVED** [review]
LGTM

**Comment by sandeep-krishnamurthy** [comment]
Merging this as it helps automating an important part of website building and all the requested changes by @marcoabreu are addressed.
@marcoabreu - Thank you for you review and inputs. Since we were unable to reach you, and I see all your comments were addressed by @waytrue17 merging this PR.
Please feel free to comment later, if there are any open items and we can cover it in another PR.

**Comment by sandeep-krishnamurthy** [comment]
@waytrue17 / @mseth10 / @ys2843 => Should this be backported to 1.X branch please?

**Comment by mseth10** [comment]
Yes, this needs to be backported to 1.x and master. @waytrue17 can you create those PRs?

**Comment by waytrue17** [comment]
> Yes, this needs to be backported to 1.x and master. @waytrue17 can you create those PRs?

Sure, I'll backport it to 1.x and master

**Comment by ys2843** [comment]
@waytrue17 It doesn't need to be port to master, only to branches that will be built as static artifact.

**Comment by mseth10** [comment]
Hey @ys2843 , I thought we will need it for any release going forward, be it 1.9 or 2.0. How do we plan to build static artifacts for 2.0 once that branch is cut?

**Pull request merged**
